miner/miner.go

```diff
@@ -2,29 +2,98 @@ package miner
 
 import (
 	"math"
+	"sync"
 	"time"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/ubclaunchpad/cumulus/blockchain"
 	"github.com/ubclaunchpad/cumulus/consensus"
 )
 
+// currentlyMining is a flag to control the miner.
+var currentlyMining bool
+
+// currentlyMiningLock is a read/write lock to change the Mining flag.
+var currentlyMiningLock sync.RWMutex
+
+const (
+	// MiningSuccessful is returned when the miner mines a block.
+	MiningSuccessful = iota
+	// MiningNeverStarted is returned when the block header is invalid.
+	MiningNeverStarted = iota
+	// MiningHalted is returned when the app halts the miner.
+	MiningHalted = iota
+)
+
+// MiningResult contains the result of the mining operation.
+type MiningResult struct {
+	Complete bool
+	Info     int
+}
+
+// RestartMiner restarts the miner with a new block.
+func RestartMiner(bc *blockchain.BlockChain, b *blockchain.Block) {
+	StopMining()
+	Mine(bc, b)
+}
+
 // Mine continuously increases the nonce and tries to verify the proof of work
 // until the puzzle is solved
-func Mine(bc *blockchain.BlockChain, b *blockchain.Block) bool {
+// TODO: Make Mine take an interface with a callback as an arguement.
+func Mine(bc *blockchain.BlockChain, b *blockchain.Block) *MiningResult {
+	setStart()
 	if valid, _ := bc.ValidBlock(b); !valid {
 		log.Error("Invalid block")
```

Comment:
I know you added this earlier but we should probably add more detail here. Generally when we surface errors we want to know what package they're from and what action what being taken when the error occurred.
In this case something like "Mine failed, miner received an invalid block".